crates/stages/src/stages/execution.rs

```diff
@@ -163,6 +181,33 @@ impl<EF: ExecutorFactory> ExecutionStage<EF> {
             done,
         })
     }
+
+    /// Adjusts the prune modes related to changesets.
+    ///
+    /// This function verifies whether the MerkleStage or Hashing stages will run from scratch.
+    /// If at least one stage isn't starting anew, it implies that pruning of changesets cannot
+    /// occur. This is determined by checking the highest clean threshold
+    /// (`self.external_clean_threshold`) across the stages.
+    ///
+    /// Given that `start_block` changes with each checkpoint, it's necessary to inspect
+    /// [`tables::AccountsTrie`] and [`tables::HashedAccount`] to ensure that the stages haven't
+    /// been previously executed.
+    fn adjust_prune_modes<DB: Database>(
+        &mut self,
+        provider: &DatabaseProviderRW<'_, &DB>,
+        start_block: u64,
+        max_block: u64,
+    ) -> Result<PruneModes, StageError> {
+        let mut prune_modes = self.prune_modes;
+        if !(max_block - start_block > self.external_clean_threshold ||
+            (provider.tx_ref().cursor_read::<tables::AccountsTrie>()?.first()?.is_none() &&
+                provider.tx_ref().cursor_read::<tables::HashedAccount>()?.first()?.is_none()))
```

Comment:
ultranit: i think `provider.tx_ref().entries::<T>()? == 0` will be cheaper here, as it doesn't require initializing a cursor and going through database

Reply:
i had that initially, but i was unsure on how it works underneath